contrib/win32/win32compat/ssh-agent/keyagent-request.c

```diff
@@ -367,6 +541,173 @@ process_remove_all(struct sshbuf* request, struct sshbuf* response, struct agent
+			provider, strerror(errno));
+		request_invalid = 1;
+		goto done;
+	}
+
+	// Remove 'drive root' if exists
+	if (canonical_provider[0] == '/')
+		memmove(canonical_provider, canonical_provider + 1, strlen(canonical_provider));
+	if (get_user_root(con, &user_root) != 0 ||
+		is_reg_sub_key_exists(user_root, SSH_PKCS11_PROVIDERS_ROOT, canonical_provider))
+		goto done;
+
+	pkcs11_init(0);
+
+	count = pkcs11_add_provider(canonical_provider, pin, &keys, NULL);
+	if (count <= 0) {
+		debug("failed to add key to store");
+		goto done;
+	}
+	for (i = 0; i < count; i++) {
+		key = keys[i];
+		memset(&sa, 0, sizeof(SECURITY_ATTRIBUTES));
+		sa.nLength = sizeof(sa);
+		if ((!ConvertStringSecurityDescriptorToSecurityDescriptorW(REG_KEY_SDDL, SDDL_REVISION_1, &sa.lpSecurityDescriptor, &sa.nLength)) ||
+			sshkey_to_blob(key, &pubkey_blob, &pubkey_blob_len) != 0 ||
+			((thumbprint = sshkey_fingerprint(key, SSH_FP_HASH_DEFAULT, SSH_FP_DEFAULT)) == NULL) ||
+			RegCreateKeyExW(user_root, SSH_KEYS_ROOT, 0, 0, 0, KEY_WRITE | KEY_WOW64_64KEY, &sa, &reg, NULL) != 0 ||
+			RegCreateKeyExA(reg, thumbprint, 0, 0, 0, KEY_WRITE | KEY_WOW64_64KEY, &sa, &sub, NULL) != 0 ||
+			RegSetValueExW(sub, NULL, 0, REG_BINARY, pubkey_blob, (DWORD)pubkey_blob_len) != 0 ||
+			RegSetValueExW(sub, L"pub", 0, REG_BINARY, pubkey_blob, (DWORD)pubkey_blob_len) != 0 ||
+			RegSetValueExW(sub, L"type", 0, REG_DWORD, (BYTE*)&key->type, 4) != 0 ||
+			RegSetValueExW(sub, L"comment", 0, REG_BINARY, canonical_provider, (DWORD)strlen(canonical_provider)) != 0) {
+			debug("failed to add key to store");
+			goto done;
+		}
+	}
+
+	debug("added smartcard keys to store");
+
+	memset(&sa, 0, sizeof(SECURITY_ATTRIBUTES));
+	sa.nLength = sizeof(sa);
+	if ((!ConvertStringSecurityDescriptorToSecurityDescriptorW(REG_KEY_SDDL, SDDL_REVISION_1, &sa.lpSecurityDescriptor, &sa.nLength)) ||
+		convert_blob(con, pin, (DWORD)pin_len, &epin, (DWORD*)&epin_len, 1) != 0 ||
+		get_user_root(con, &user_root) != 0 ||
+		RegCreateKeyExW(user_root, SSH_PKCS11_PROVIDERS_ROOT, 0, 0, 0, KEY_WRITE | KEY_WOW64_64KEY, &sa, &reg, NULL) != 0 ||
+		RegCreateKeyExA(reg, canonical_provider, 0, 0, 0, KEY_WRITE | KEY_WOW64_64KEY, &sa, &sub, NULL) != 0 ||
+		RegSetValueExW(sub, L"provider", 0, REG_BINARY, canonical_provider, (DWORD)strlen(canonical_provider)) != 0 ||
+		RegSetValueExW(sub, L"pin", 0, REG_BINARY, epin, (DWORD)epin_len) != 0) {
+		debug("failed to add pkcs11 provider to store");
+		goto done;
+	}
+
+	debug("added pkcs11 provider to store");
+	success = 1;
+done:
+	r = 0;
+	if (request_invalid)
+		r = -1;
+	else if (sshbuf_put_u8(response, success ? SSH_AGENT_SUCCESS : SSH_AGENT_FAILURE) != 0)
+		r = -1;
+
+	/* delete created reg keys if not succeeded*/
+	if ((success == 0) && reg) {
+		if (thumbprint)
+			RegDeleteKeyExA(reg, thumbprint, KEY_WOW64_64KEY, 0);
+		if (canonical_provider)
+			RegDeleteKeyExA(reg, canonical_provider, KEY_WOW64_64KEY, 0);
+	}
+	pkcs11_terminate();
```

Comment:
Here we risk calling `pkcs11_terminate()` without calling `pkcs11_init()` first. While it seems harmless, it definitely feels like an anti-pattern, since it does not appear to happen anywhere else.

Reply:
I've bumped `pkcs11_init()` to the beginning of the function to address that and comply with the pattern. Thanks.